### PR TITLE
Nc 1815

### PR DESCRIPTION
**Pull request by shemnon**
## PR description

as recommended in #231 alter the API so that the
subclasses always return their list from executePeerTask in the
`CompletableFuture` and reset the retry in `#executeTask()` when any
non-empty list is returned.

### CODE REVIEW

File: ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/manager/AbstractRetryingPeerTask.java [hunk 1]
```diff
@@ -18,23 +18,20 @@
...
- * <p>As an additional semantic subclasses may call {@link #resetRetryCounter} so that if they have
- * partial success they can reset the retry counter and only count zero progress retries against the
- * exception limit. If this facility is used only consecutive zero progress retries count against
- * the maximum retries limit.
- *
- * @param <T> The type of the CompletableFuture this task will return.
+ * @param <T> The type as a typed list that the peer task can get partial or full results in.
  */
-public abstract class AbstractRetryingPeerTask<T> extends AbstractEthTask<T> {
+public abstract class AbstractRetryingPeerTask<T extends List<?>> extends AbstractEthTask<T> {
```
Comment (mbaxter) [inline ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/manager/AbstractRetryingPeerTask.java]: Does it make sense to use the more generic `Collection` here?

File: ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/manager/AbstractRetryingPeerTask.java [hunk 2]
```diff
@@ -68,12 +65,16 @@ protected void executeTask() {
               if (error != null) {
                 handleTaskError(error);
               } else {
+                // If we get a partial success reset the retry counter.
+                if (peerResult.size() > 0) {
+                  retryCount = 0;
```
Comment (mbaxter) [inline ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/manager/AbstractRetryingPeerTask.java]: You can remove `resetRetryCounter()` now.

Comment (shemnon) [reply ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/manager/AbstractRetryingPeerTask.java]: Github merge restored it.

Comment (shemnon) [reply ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/manager/AbstractRetryingPeerTask.java]: note to self: never use github web merge again.

**Pull request merged**